docs/design/2020-05-08-cluster-index.md

```diff
@@ -0,0 +1,254 @@
+
+### Planner
+
+#### Handle Primary Key Index differently
+For common handle tables, treat the primary index access path is table path instead of index path, builds TableScan plan instead of IndexLookUp plan.
+
+#### IndexScan Schema
+
+If any of the primary key columns are used, or it is in an IndexLookUp executor, the index scan schema should be:
+
+`{index columns},{primary key columns}`
+
+There is no extra handle column for common handle tables.
+
+If only index columns are used, the schema should be:
+
+`{index columns}`
+
+For example:
+```sql
+create table t (a int, b int, c int, d int, e int, primary key (a, b), index c_d (c, d));
+```
+The schema of index scan on `c_d` should be `c, d, a, b` or `c, d` depends on whether any primary key column is used.
+
+So if a query doesn't reference column `e`, the index `c_d` is a covering index, we can build an IndexScan plan
+instead of an IndexLookUp plan.
+
+### Coprocessor
+Besides codec code, several coprocessor executors need to be changed to support common handle.
+
+#### IndexScan
+If the IndexScan schema doesn't contain primary key columns, the coprocessor logic is unchanged.
+
+If the IndexScan schema contains primary key columns, the coprocessor should decode the common handle
+to primary key column values.
+
+The encoded column value can be obtained by the `EncodedCol` method of the handle.
+
+#### Analyze Column
+If the primary key is multi-column, CMSketch needs to insert prefix column values of the primary key. 
+
+#### Fast Analyze
+Since the handle is not integer anymore, we cannot randomly generate the sample key at a given position.
+So we need to add a coprocessor request type that samples the key/value pair during the scan and return the statistics result.
+
+### Executors
+Several executors need to be changed to support common handle
+
+#### Admin Executors
+The `admin check/recover table/index` executors heavily depend on the assumption that handle is int64 handle, 
+it needs to do major changes to support common handle.
+
+#### PointGet/BatchPointGet
+We need to decode the common handle to chunk. 
+
+#### Insert
+We don't need to encode primary columns into the row value.
+
+#### Update
+If any of the primary key columns changed, we need to delete the row and insert the row.
+
+#### SplitTableRegion
+We need to redesign this executor since we cannot find a middle key based on the region start/end key. 
+
+## Rationale
+
+The clustered index feature would remove the extra read/write cost for tables with a non-single-integer-column primary key.
+An experiment on the tpcC workload by compressing multi-column PK into one integer PK shows a 33% improvement. 
+
+## Compatibility
+
+This feature doesn't break the backward compatibility.
+
+We add a field `IsCommonHandle` in the `model.TableInfo` and a global variable `tidb_enable_clustered_index`.
+
+When creating a new table, set this field to true if the global variable `tidb_enable_clustered_index` is set to `1`.
```

Comment:
The old table doesn't change even if we alter it, `IsComonHandle` is false in the old `model.TableInfo`.

Reply:
The attempt is not easy to be supported, right? It is basically the same with `ALTER ... PRIMARY_KEY` on pk_is_handle table.

However, we've a plan to support it in the near future, so it would be possible to support it.

This should be useful when users encountered a hotspot write when `tidb_enable_clustered_index=1` and they don't want to create a new table without clustered index.